### PR TITLE
Detect `std` by checking if the crate defines `#[lang = "start"]` rather than string comparison

### LINKED ISSUE #1821: `extern crate std;` in a `#![no_std]` crate confuses Miri unless it's in the root module
This program:
```rs
#![no_std]

fn main() {
    extern crate std;
}
```
Causes this error in Miri ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=81a82dcc0740bd324db6c6ffd1fe97d3)):
```
error: unsupported operation: can't call foreign function: signal
   --> /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/unix/mod.rs:119:17
    |
119 |         assert!(signal(libc::SIGPIPE, libc::SIG_IGN) != libc::SIG_ERR);
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't call foreign function: signal
    |
    = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
            
    = note: inside `main::std::sys::unix::init::reset_sigpipe` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/unix/mod.rs:119:17
    = note: inside `main::std::sys::unix::init` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/unix/mod.rs:62:5
    = note: inside `main::std::sys_common::rt::init` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/rt.rs:16:9
    = note: inside `main::std::rt::lang_start_internal` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:32:14
    = note: inside `main::std::rt::lang_start::<()>` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:48:5
```

But the program runs fine if [`extern crate std;` is in the root module](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=3b92981a84bfe92a3c8ce916db46b5a0) or [the crate is not `#![no_std]`](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=6b04b979a68a49ad51fdc59232c35c15).

I encountered this issue when I was trying to access `std` by using `extern crate std;` in a `#[test]` of a `#![no_std]` crate.

I guess that this is because the stack frame is `main::std::...` in this case, while this code (cc #1263) only checks for `std::...`:
https://github.com/rust-lang/miri/blob/453affaaa1762a065d2857970b8333017211208c/src/shims/posix/foreign_items.rs#L504

### PR DESCRIPTION
**Pull request by hyd-dev**
I also considered to compare the crate name with `sym::std`, but it's easy to name any crate `std` by using `--crate-name std`, so I don't think that is robust enough.

Note that this only checks the crate, it does not check whether the call is in `sys::unix` or `sys::windows`, unlike the previous implementation, but I think it's already robust enough.

Fixes #1821.

### CODE REVIEW

File: src/helpers.rs [hunk 1]
```diff
@@ -616,6 +616,12 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
         Ok(wchars)
     }
+
+    fn in_std(&self) -> bool {
+        let this = self.eval_context_ref();
+        this.tcx.def_path(this.frame().instance.def_id()).krate
+            == this.tcx.def_path(this.tcx.lang_items().start_fn().unwrap()).krate
```
Comment (bjorn3) [inline src/helpers.rs:623]: There may not be a `#[lang = "start"]` when using `#![no_std]` in combination with `#[start]`, so please don't unwrap.

Comment (hyd-dev) [reply src/helpers.rs:623]: There is already an `unwrap()` in `create_ecx()` :thinking:
https://github.com/rust-lang/miri/blob/ef998308805748289e46bce394a5d539b12b0118/src/eval.rs#L113

**Comment by RalfJung** [comment]
Note that the check does not have to be extremely robust. If someone wants to go out of their way to use these incomplete Miri shims, I'm fine with that, but they don't get to complain about any bugs they encounter. ;)

So I'm in favor of your code changes, but I am worried about the test size blowup.^^ `tests/run-pass/std_only_foreign_function.rs` is *not* a supported use-case, so we shouldn't test it IMO. For the compile-fail test, I think having just one is enough -- just to make sure that `in_std` does not always return `true`. So I'm in favor of also removing `tests/compile-fail/unsupported_get_process_heap.rs`.

File: src/helpers.rs [hunk 2]
```diff
@@ -616,6 +616,14 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
         Ok(wchars)
     }
+
+    fn in_std(&self) -> bool {
```
Comment (RalfJung) [inline src/helpers.rs:620]: The name is not very self-explaining... *what* is in std?
Please at least add a doc-comment. Maybe there's a more self-explaining name?

File: src/helpers.rs (hunk 1 above)
Comment (RalfJung) [reply src/helpers.rs]: > using #![no_std] in combination with #[start]

That sounds like a test case we should have, then. ;)

File: src/helpers.rs (hunk 2 above)
Comment (hyd-dev) [reply src/helpers.rs]: This calls `this.frame()`... so `frame_in_std`?

File: src/helpers.rs (hunk 1 above)
Comment (hyd-dev) [reply src/helpers.rs]: > using `#![no_std]` in combination with `#[start]`

Note that it's currently broken even before the `unwrap()`:
```
error: internal compiler error: src/tools/miri/src/eval.rs:110:9: main function must not take any arguments

thread 'rustc' panicked at 'Box<Any>', compiler/rustc_errors/src/lib.rs:1007:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.54.0-nightly (c79419af0 2021-06-04) running on x86_64-unknown-linux-gnu
```
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=b867b78dd787c573bdfb44c88fa856dc
https://github.com/rust-lang/miri/blob/ef998308805748289e46bce394a5d539b12b0118/src/eval.rs#L110

**Comment by hyd-dev** [comment]
> `tests/run-pass/std_only_foreign_function.rs` is _not_ a supported use-case, so we shouldn't test it IMO.

I added it to test that `in_std()` does not always return `false`, which is not *explicitly* tested in other tests.

File: src/helpers.rs (hunk 1 above)
Comment (RalfJung) [reply src/helpers.rs]: Hm, the start item in Rust has no arguments so I am not sure if it makes sense to support other start items that do.

But this sounds like a separate discussion then, could you open an issue?

File: src/helpers.rs (hunk 2 above)
Comment (RalfJung) [reply src/helpers.rs]: Yeah, makes sense. :)

File: tests/compile-fail/unsupported_signal.rs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+//! `signal()` is special on Linux and macOS that it's only supported within libstd.
```
Comment (RalfJung) [inline tests/compile-fail/unsupported_signal.rs:1]:
```suggestion
//! `signal()` is special on Linux and macOS that it's only supported within libstd.
//! The implementation is not complete enough to permit user code to call it.
```

**Comment by RalfJung** [comment]
Thanks @hyd-dev :)
@bors r+

**Comment by RalfJung** [comment]
> (I have removed it now, but) I added it to test that in_std() does not always return false, which is not explicitly tested in other tests.

Fair -- however, it is implicitly tested in every single test, which IMO is good enough.

**Pull request merged**